lib/foreman/renderer/scope/macros/helpers.rb

```diff
@@ -11,6 +11,33 @@ module Helpers
             sections only: %w[all]
           end
 
+          apipie :method, "Returns counter increased by given value (as String) stored in a hostgroup param named 'sequence_num_xxx'" do
+            required :hostgroup_name, String, "hostgroup title to store in (e.g. RHEL7/base)"
+            optional :increment, Integer, "integer increment, can be negative too", default: 1
+            optional :name, String, "name of the sequence (prefix for hostgroup parameter)", default: 'default'
+            optional :prefix, Integer, "leading zeroes", default: 1
+            returns String, "String with incremented value or empty string if hostgroup not found"
+            example 'sequence_hostgroup_param_next("myHG", 10) #=> 10'
+            example 'sequence_hostgroup_param_next("myHG", 2, "kiosk", 5) #=> 00002'
+          end
+          def sequence_hostgroup_param_next(hostgroup_name, increment = 1, name = "default", prefix = 1)
+            hostgroup = Hostgroup.find_by_title(hostgroup_name)
+            return '' unless hostgroup
+            result = 0
+            Parameter.transaction do
+              ActiveRecord::Base.connection.execute('LOCK parameters IN ACCESS EXCLUSIVE MODE')
```

Comment:
Could this lock cause issues for other users/templates attempting to modify parameters at the same time?

Reply:
That is the purpose of table exclusive lock, I can use row-level locking the problem is this isn't just an update, there's `find_or_create` and I could not find a good solution to that. But after some more research I have stumbled upon advisory locking: https://www.kostolansky.sk/posts/postgresql-advisory-locks/ and I implemented this solution actually.